src/lib/System/IO/Streams/Heartbeat.hs

```diff
@@ -1,75 +1,110 @@
+import Control.Concurrent.Async (async, cancel, link)
+import Control.Exception (Exception, throw)
+import Control.Monad (forever, void)
+import Data.Foldable (traverse_)
+import Data.IORef (IORef, atomicModifyIORef', newIORef, readIORef, writeIORef)
+import Data.Maybe (fromJust)
+import Data.Time.Clock (DiffTime, UTCTime, diffTimeToPicoseconds, diffUTCTime, getCurrentTime)
+import System.IO.Streams (InputStream, OutputStream)
+import qualified System.IO.Streams as Streams
+
+
+-- | Send a message 'a' if nothing has been written on the stream for some
+-- interval of time.
+--
 -- Writing 'Nothing' to this 'OutputStream' is required for proper cleanup.
-heartbeatOutputStream :: DiffTime -- ^ Heartbeat interval
-                      -> a        -- ^ Heartbeat message
-                      -> OutputStream a -> IO (OutputStream a)
+--
+-- Also returns an 'IO' action that can be used to dynamically update the
+-- heartbeat interval, where 'Nothing' indicates that heartbeating should be
+-- disabled.
+heartbeatOutputStream ::
+    -- | Heartbeat interval
+    Maybe DiffTime ->
+    -- | Heartbeat message
+    a ->
+    OutputStream a ->
+    IO (OutputStream a, Maybe DiffTime -> IO ())
 heartbeatOutputStream interval msg os = do
     t <- newIORef =<< getCurrentTime
-    writeAsync <- async $ delayInterval >> forever (writeHeartbeat t)
+    intervalRef <- newIORef interval
+    writeAsync <- async $ delayInterval >> forever (writeHeartbeat t intervalRef)
     link writeAsync
-    Streams.makeOutputStream (resetHeartbeat t writeAsync)
+    (,updateHeartbeatInterval intervalRef) <$> Streams.makeOutputStream (resetHeartbeat t writeAsync)
   where
     delayInterval = delayDiffTime interval
 
-    writeHeartbeat t = do
+    writeHeartbeat t intervalRef = do
         !now <- getCurrentTime
-        (!timeTilHeartbeat, !triggerHeartbeat) <- atomicModifyIORef' t (heartbeatTime interval now)
+        int <- readIORef intervalRef
+        (!timeTilHeartbeat, !triggerHeartbeat) <- atomicModifyIORef' t (heartbeatTime int now)
 
         if triggerHeartbeat
             then Streams.write (Just msg) os >> delayInterval
             else delayDiffTime timeTilHeartbeat
 
-    resetHeartbeat t _ x@(Just _)       = Streams.write x os >> getCurrentTime >>= writeIORef t
+    resetHeartbeat t _ x@(Just _) = Streams.write x os >> getCurrentTime >>= writeIORef t
     resetHeartbeat _ writeAsync Nothing = Streams.write Nothing os >> cancel writeAsync
 
 
 -- | Exception to kill the heartbeat monitoring thread
 -- Heartbeat Exceptions carry the grace period, ie. the last time a message was received
-data HeartbeatException = MissedHeartbeat DiffTime deriving (Show, Eq)
+newtype HeartbeatException = MissedHeartbeat DiffTime deriving (Show, Eq)
+
+
 instance Exception HeartbeatException
 
 
+updateHeartbeatInterval :: IORef (Maybe DiffTime) -> (Maybe DiffTime -> IO ())
+updateHeartbeatInterval ref int = void $ atomicModifyIORef' ref (const (int, int))
```

Comment:
I don't think you need the void. You can just return `()` inside the `atomicModifyIORef'`.